bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/MetadataDrivers.java

```diff
@@ -0,0 +1,290 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.bookkeeper.meta;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Sets;
+import java.net.URI;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.ToString;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.bookkeeper.util.ReflectionUtils;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * A driver manager for managing a set of metadata drivers.
+ *
+ */
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+@Slf4j
+public final class MetadataDrivers {
+
+    static final String ZK_CLIENT_DRIVER_CLASS = "org.apache.bookkeeper.meta.zk.ZKMetadataClientDriver";
+    static final String ZK_BOOKIE_DRIVER_CLASS = "org.apache.bookkeeper.meta.zk.ZKMetadataBookieDriver";
+    static final String BK_METADATA_CLIENT_DRIVERS_PROPERTY = "bookkeeper.metadata.client.drivers";
+    static final String BK_METADATA_BOOKIE_DRIVERS_PROPERTY = "bookkeeper.metadata.bookie.drivers";
+
+    @ToString
+    static class MetadataClientDriverInfo {
+
+        final Class<? extends MetadataClientDriver> driverClass;
+        final String driverClassName;
+
+        MetadataClientDriverInfo(Class<? extends MetadataClientDriver> driverClass) {
+            this.driverClass = driverClass;
+            this.driverClassName = driverClass.getName();
+        }
+
+    }
+
+    @ToString
+    static class MetadataBookieDriverInfo {
+
+        final Class<? extends MetadataBookieDriver> driverClass;
+        final String driverClassName;
+
+        MetadataBookieDriverInfo(Class<? extends MetadataBookieDriver> driverClass) {
+            this.driverClass = driverClass;
+            this.driverClassName = driverClass.getName();
+        }
+
+    }
+
+    @Getter(AccessLevel.PACKAGE)
+    private static final ConcurrentMap<String, MetadataClientDriverInfo> clientDrivers;
+    @Getter(AccessLevel.PACKAGE)
```

Comment:
I can't find any code to release resources/unload the drivers.
Can you please give me a pointer?

Reply:
I think I forgot to add the unload interfaces. However since we also use this pattern on dlog namespace driver, I don't see a real need to unload drivers. 

Also This only caches the class information when the drivers are included in the classpath, it doesn't contain any real resources that we are using.

Reply:
I would prefer to have a way to cleanup things, think about webapps, in this case the webapplication need a way to cleanup things.
Or maybe we can have an *unregister* method to unload the driver explicitly

Reply:
I can add an unregister method. But my point is it is never used, because in current approach, we only keep the mapping between scheme and the class. We don’t instantiate objects until resolving service uri. So there is really no place for you to unregister a “Class”. This is different from web apps. 

Franckly speaking, you only will have one implementation in your classpath and only use that implemention all the time. There is really no place or point for you to unregister.